libtextsecure/message_receiver.js

```diff
@@ -1286,14 +1286,81 @@ MessageReceiver.prototype.extend({
       this.processDecrypted(envelope, msg).then(async message => {
         const groupId = message.group && message.group.id;
         const isBlocked = this.isGroupBlocked(groupId);
-        const isMe = envelope.source === textsecure.storage.user.getNumber();
+        const ourPubKey = textsecure.storage.user.getNumber();
+        const isMe = envelope.source === ourPubKey;
         const conversation = window.ConversationController.get(envelope.source);
         const isLeavingGroup = Boolean(
           message.group &&
             message.group.type === textsecure.protobuf.GroupContext.Type.QUIT
         );
         const friendRequest =
           envelope.type === textsecure.protobuf.Envelope.Type.FRIEND_REQUEST;
+        const { UNPAIRING_REQUEST } = textsecure.protobuf.DataMessage.Flags;
+        // eslint-disable-next-line no-bitwise
+        const isUnpairingRequest = Boolean(message.flags & UNPAIRING_REQUEST);
+
+        if (isUnpairingRequest) {
+          // TODO: move high-level pairing logic to libloki.multidevice.xx
+
+          const unpairingRequestIsLegit = async () => {
+            const isSecondary = textsecure.storage.get('isSecondaryDevice');
+            if (!isSecondary) {
+              return false;
+            }
+            const primaryPubKey = window.storage.get('primaryDevicePubKey');
+            // TODO: allow unpairing from any paired device?
+            if (envelope.source !== primaryPubKey) {
```

Comment:
Should we also check that the message is not a friend request? I could send a friend request to another secondary device with the primary device public key as the envelope source. (unless signal already verifies that somewhere)

Reply:
Good question... I assume that the message is decrypted _using_ the envelope source, so in theory if we got this far we should be able to trust that the envelope source is actually the sender. I'm not 100% sure thought...
Luckily we need the primary device's device mapping to reflect the unpairing as well to trigger the actual wiping of the device :)
Anyways, I guess it wouldn't hurt to do 
```js
if (!friendRequest && isUnpairingRequest) {
```
Would that be better?

Reply:
Hmm that's a good point, it should be fine to leave it as is :)